modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java

```diff
@@ -19,371 +19,302 @@
+                Method method = struct.methods.get(name);
+                if (method != null) {
+                    return method.handle;
+                }
+            }
 
-        if (types.size() != arguments.length) {
-            throw new IllegalArgumentException("When dynamically calling [" + name + "] from class " +
-                    "[" + owner.getClass() + "] expected [" + types.size() + "] arguments," +
-                    " but found [" + arguments.length + "].");
-        }
+            for (Class<?> iface : clazz.getInterfaces()) {
+                struct = definition.runtimeMap.get(iface);
 
-        try {
-            for (int count = 0; count < arguments.length; ++count) {
-                if (typesafe[count]) {
-                    parameters[count + 1] = arguments[count];
-                } else {
-                    final Transform transform = getTransform(arguments[count].getClass(), types.get(count).clazz, definition);
-                    parameters[count + 1] = transform == null ? arguments[count] : transform.method.handle.invoke(arguments[count]);
+                if (struct != null) {
+                    Method method = struct.methods.get(name);
+                    if (method != null) {
+                        return method.handle;
+                    }
                 }
             }
-
-            return handle.invokeWithArguments(parameters);
-        } catch (Throwable throwable) {
-            throw new IllegalArgumentException("Error invoking method [" + name + "] " +
-                    "with owner class [" + owner.getClass().getCanonicalName() + "].", throwable);
         }
-    }
 
-    @SuppressWarnings({ "unchecked", "rawtypes" })
-    public static void fieldStore(final Object owner, Object value, final String name,
-                                  final Definition definition, final boolean typesafe) {
-        final Field field = getField(owner, name, definition);
-        MethodHandle handle = null;
+        // no matching methods in whitelist found
+        throw new IllegalArgumentException("Unable to find dynamic method [" + name + "] " +
+                                           "for class [" + receiverClass.getCanonicalName() + "].");
+    }
 
-        if (field == null) {
-            final String set = "set" + Character.toUpperCase(name.charAt(0)) + name.substring(1);
-            final Method method = getMethod(owner, set, definition);
+    /** pointer to Array.getLength(Object) */
+    private static final MethodHandle ARRAY_LENGTH;
+    /** pointer to Map.get(Object) */
+    private static final MethodHandle MAP_GET;
+    /** pointer to Map.put(Object,Object) */
+    private static final MethodHandle MAP_PUT;
+    /** pointer to List.get(int) */
+    private static final MethodHandle LIST_GET;
+    /** pointer to List.set(int,Object) */
+    private static final MethodHandle LIST_SET;
+    static {
+        Lookup lookup = MethodHandles.lookup();
```

Comment:
This should better be a MethodHandles.publicLookup(), as we only lookup public methods from public classes. This is not critical here, as its a static part only, but for "correctness".